### ISSUE TITLE
[@typescript-eslint/no-unused-vars] false positive

### ISSUE THREAD

**Issue by benkeil** [typescript-eslint/typescript-eslint#3633]
```js
{
  rules: {
    'no-unused-vars': [
      'off',
      {
        args: 'after-used',
        argsIgnorePattern: '^_',
        ignoreRestSiblings: true,
        caughtErrors: 'none',
      },
    ],
    '@typescript-eslint/no-unused-vars': [
      'error',
      {
        args: 'after-used',
        argsIgnorePattern: '^_',
        ignoreRestSiblings: true,
        caughtErrors: 'none',
      },
    ],
  }
}
```

```TS
const array: string[] = [];
const [message, ...params] = array;
console.log(message);
const [message1, ..._params] = array;
console.log(message1);
```

tsconfig.json
```
{
  "compilerOptions": {
    "alwaysStrict": true,
    "baseUrl": "lib",
    "declaration": false,
    "downlevelIteration": true,
    "emitDecoratorMetadata": true,
    "esModuleInterop": true,
    "experimentalDecorators": true,
    "lib": [
      "ESNext",
      "ES2015",
      "DOM"
    ],
    "module": "CommonJS",
    "moduleResolution": "node",
    "noUnusedLocals": false,
    "noImplicitAny": false,
    "outDir": "dist",
    "preserveSymlinks": true,
    "resolveJsonModule": true,
    "sourceMap": true,
    "strict": true,
    "target": "ES6"
  },
  "exclude": [
    "node_modules",
    "dist"
  ],
  "include": [
    "lib",
    "index.js",
    "react.js",
    "tests"
  ]
}
```

**Expected Result**

`_params` and `params` is valid in this case, because of `ignoreRestSiblings: true` and also `argsIgnorePattern: true`

**Actual Result**

```
ESLint: '_params' is assigned a value but never used.(@typescript-eslint/no-unused-vars)
```

**Additional Info**

no

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.3` |
| `@typescript-eslint/parser`        | `4.28.3` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.30.0` |
| `node`                             | `v14.15.4` |

**Comment by bradzacher** [comment]
This is working as expected:
1) `argsIgnorePattern` applies to ***arguments***. These are variables.
2) `ignoreRestSiblings` applies to the siblings of a rest operator, not the rest operator itself.

https://eslint.org/docs/rules/no-unused-vars

**Comment by benkeil** [comment]
Then extend the configuration of the rule. It covers only a small amount of cases.

**Comment by bradzacher** [comment]
What do you mean? Have you read the docs? There is an option specifically for ignoring variables.

---

Also including an unused rest arg is also literally useless. It doesn't change the value of any of its siblings.

`...y` means "put all other elements in an array and assign it to `y`"

`const [x, ...unused] = arr;`
Is exactly the same as
`const [x] = arr;`

This is why it's not ignored by the `ignoreRestSiblings` option.

**Comment by benkeil** [comment]
omg, you'r right... 🤭
